### DISCUSSION TITLE
Conman and Conman CLI

### DISCUSSION THREAD

**Discussion by tybruno** [carlmontanari/scrapli#149] [Show and tell]
Hello @carlmontanari,

I know you are probably busy but I was wondering if you could take a look at a library I am building and its Python CLI counterpart. They both use Scrapli as the connection handler. I would like general feedback. The Testing, code, and documentation are not complete but I would just like to make sure I am on the right track before I get too far hahaha 😆. I am interested in opinions regarding design and structure.

Feel free to give me feedback of any kind. I am not 100% if I like the name but it is all I could think of at the moment. 

Also, if there is anything in my code that you think would be better placed into Scrapli, I could always migrate it 😄.

I hope to make the project's documentation and structure as similar to Scrapli as possible so that users can transition between Scrapli docs and Conman docs without any difficulty. 

## [Conman](https://github.com/tybruno/conman) (Connections Manager):
The goal of Conman is to make running commands on multiple devices even easier. I also want to be able to save common "Action Runners" / "Runbooks" for devices. 

These action runners can be extended to perform pre-built actions on devices. For example, let's say we have an action runner named "ResetDevices" where it has the actions needed to reset the given devices. 

### Responsibilities:
-  Generate Handlers for multiple connections
-  Run actions on multiple devices
-  Handle the action results

### TODO:
- Add Async Support
- Finish Docs
- Finish Testing

##### What advice am I looking for?
- project structure: How should I structure the project so that developers can upload their own Actions, Action handlers, and Runners? Should I break them into their own Repo's? 

## [Conman CLI](https://github.com/tybruno/conman_cli)
The goal of conman CLI is to provide the networking engineers an easy CLI for them to run multiple commands to multiple devices via CLI.

This is going to be the step before I make a web app version of this where people can install and run a web app from their own internal structure.

### Responsibilities
- Run a batch of CLI commands on multiple devices (this should be able to change privilege levels based on cli input) and display the output
- CLI commands can be loaded from .txt, .xlsx, or .csv file as well as the hosts. This will allow the networking engineer to save and share batch of commands that solve problems (i.e. batch of commands to reset devices)

### TODO:
-  Add Async support
-  Add file reader that can read commands and/or hosts from .txt, .xlsx, or .csv
-  freeze with pyinstaller

Thank you again for all your help! Scrapli is an awesome tool. I am excited to extend it's capabilities 😄 

Tyler

**Comment by carlmontanari** [comment]
> I hope to make the project's documentation and structure as similar to Scrapli as possible so that users can transition between Scrapli docs and Conman docs without any difficulty.

One thing I hate about scrapli docs rn is the auto generated api docs section. I have been meaning to look at mkdocstrings for forever but haven't gotten around to it. I really like the idea/general thing we end up w/ in scrapli docs -- searchable docstrings with types and parameters and returns and all that stuff, just dont like how I shoe horned it into the docs... so if there is a better way (with mkdocstrings or other) that would be something worth investigating I think! And for sure would be something cool to bring back into scrapli if you get something cool dialed in!

> project structure: How should I structure the project so that developers can upload their own Actions, Action handlers, and Runners? Should I break them into their own Repo's?

Billion dollar question ya? lol I dont know if I have any smart thoughts on this... I've done something kinda like this in the past for a dayjob$ and it was so hard to make something that was both simple but also super flexible. In general I'd say that having one or several "templates" to extend is a smart thing -- that "template" could be a class that users extend or an object that users interact with (like they pass it hosts/functions to execute or something)... but I think you will probably be best served with just a ton of trial and error to find what feels right! One last thing Id say here -- make it easy to meaningfully test... stuff that can't be tested is a bad bingo obviously -- thats a big reason I dislike Ansible (yes molecule, yes yamllint and blah blah but none of that is comparable to pytest imo). 

This sounds like it will be pretty cool stuff, def keep us posted on your progress! 

Carl

**Reply by tybruno** [reply]
Hey @carlmontanari,

The only api docs generator that I have experience with is Sphinx.  I think this idea is worth investigating. I'll circle back to this soon.

Sorry for the late response. I was in am Object Oriented Design Principles class the past few days. I am going to do some rework on the design and then reply again when I have a better overall design. 

Thank you!

**Reply by tybruno** [reply]
Hey @carlmontanari,

i did a significant rework of the project. I also renamed it to "[Scraplis](https://github.com/tybruno/scraplis)" since the whole goal is to make it easier to work with multiple Scrapli handlers.  I am also in the works of finishing the "[Scraplis_cli](https://github.com/tybruno/scraplis_cli)". I will post that on here when it is ready for a look.

Please let me know what you think. Let me know if you think the name is appropriate or not.

**Reply by carlmontanari** [reply]
Heyo! Nice, that CLI should be nice and handy for some folks I think!

To be honest I don't love the name -- I'm flattered and I think its cool, but my concern would be that is one typo away from being very confusing for users looking for either `scrapli` or `scraplis`. You are of course more than welcome to keep it I totally won't be upset/offended or anything, that was just my first thought when I saw it so figured I'd share.

Carl

**Reply by tybruno** [reply]
Hey @carlmontanari,

The foundation for "[Scraplis_cli](https://github.com/tybruno/scraplis_cli)" is located here.

Thank you for the feedback :) I was also thinking the same thing. Maybe something like "multi_scrapli" would be better then have the cli be called "scrapli_cli" ? hahaha 

Tyler

**Reply by carlmontanari** [reply]
`multi_scrapli` sounds fitting! Could do `multi_scrapli_ctl` for the CLI to sound super fancy too! 😁
